etcdserver/api/v3rpc/watch.go

```diff
@@ -189,19 +192,25 @@ func (ws *watchServer) Watch(stream pb.Watch_WatchServer) (err error) {
 		}
 	}()
 
+	defer sws.close()
 	select {
 	case err = <-errc:
 		close(sws.ctrlStream)
 
 	case <-stream.Context().Done():
 		err = stream.Context().Err()
-		// the only server-side cancellation is noleader for now.
```

Comment:
is there a better way to decide if the cancel is from server side or from client side?

Reply:
yeah, good question I dug pretty hard but let me review again.